### PR TITLE
[Feature] RTT Control Buffer and Status

### LINKED ISSUE #48: how to get RTT buffer size?
The J-Link RTT Viewer can get channel description include buffer size,like below
channel info
Is there a API to do this? I try rtt_control,but failed.

### PR DESCRIPTION
**Pull request by hkpeprah**
### Description
This PR fleshes out the RTT support to resolve #48 and #51.  Specifically, this PR adds the ability to pass an address for the RTT control block when starting RTT, get the current RTT status (including number of up and down buffers), and get the descriptor for an RTT buffer (which returns information such as the name of the buffer, and the size of the buffer).

### Testing
A new firmware image has been added that builds in the RTT vendor code and implements a simple echo program.  The firmware reads a single byte from an RTT down buffer and writes that byte to the RTT up buffer; both on channel 0.  The associated functional test attempts to write a string over RTT on the host, then read said string back from the device.  This has been tested and verified on an evaluation board.

CC @charlesnicholson, as original RTT support adder.

### CODE REVIEW

**Comment by LeMimit** [comment]
Hello,

Do you know when this PR will be merged ?

Thanks a lot

**Comment by hkpeprah** [comment]
> Hello,
> 
> Do you know when this PR will be merged ?
> 
> Thanks a lot

Just waiting on a review.  CC @charlesnicholson @sstallion

**Comment by charlesnicholson** [comment]
It's my official position that @sstallion should get off his lazy ass and review this excellent code.

**Comment by sstallion** [comment]
Review it yourself, jerk. Oh wait you can't.

(I'll take a look in a moment)

File: tests/unit/test_structs.py [hunk 1]
```diff
@@ -400,6 +400,70 @@ def test_jlink_trace_region(self):
...
+        self.assertTrue(desc.up)
+        self.assertFalse(desc.down)
+
+        desc.Direction = 1
+        self.assertTrue(desc.down)
+        self.assertFalse(desc.up)
+
+        self.assertEqual('The Winds of Winter', desc.name)
+        self.assertEqual('JLinkRTTerminalBufDesc(Index=7, Name=The Winds of Winter)', repr(desc))
+        self.assertEqual('The Winds of Winter <Index=7, Direction=down, Size=1337>', str(desc))
```
Comment (sstallion) [inline tests/unit/test_structs.py]: One day...

File: tests/functional/firmware/k21-rtt/asm/bootloader.s [hunk 2]
```diff
@@ -0,0 +1,188 @@
+/* File: bootloader.s
```
Comment (sstallion) [inline tests/functional/firmware/k21-rtt/asm/bootloader.s]: Hah! I wasn't expecting for this to ever see the light of day. I guess what goes around comes around.

File: tests/functional/firmware/k21-rtt/src/main.c [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * Created: March 12th, 2019
+ * Modified: March 12th, 2019
+ *****************************************************************************/
+
+#include <SEGGER_RTT.h>
+
+#include <string.h>
+
+#define FOREVER  for(;;)
```
Comment (sstallion) [inline tests/functional/firmware/k21-rtt/src/main.c]: NIT: doesn't seem to be used

**Comment by charlesnicholson** [comment]
And that, dear readers, is what effective management looks like.

File: tests/functional/firmware/k21-rtt/src/main.c (hunk 3 above)
Comment (hkpeprah) [reply tests/functional/firmware/k21-rtt/src/main.c]: Woops, yah.  Copied it all from the other FW images, haha.

Comment (hkpeprah) [reply tests/functional/firmware/k21-rtt/src/main.c]: Wait, it is used below on line 48.

**Pull request merged**